### PR TITLE
Catch OverflowError

### LINKED ISSUE #4527: The "OSS-Fuzz Findings" test seems to fail on 32-bit machines
### Brief description

https://download.copr.fedorainfracloud.org/results/packit/evverx-scapy-2/fedora-40-i386/07990313-scapy/builder-live.log.gz
```python
###(167)=[failed] OSS-Fuzz Findings

>>> from io import BytesIO
>>> file = BytesIO(b"\n\r\r\n\x1c\x00\x00\x00M<+\x1a\x01\x00\x00\x00\xff\xff\xff\xff\xff\xff\xff\xff\x1c\x00\x00\x00\x01\x00\x00\x00\x14\x00\x00\x00\xe4\x00\x00\x00\x00\x00\x04\x00\x14\x00\x00\x00\x01\x00\x00\x00(\x00\x00\x00\xe4\x00\x00\x00\x00\x00\x04\x00\x02\x00\t\x00b'ens16\xb0'\x00\x00\x00\x00\x00\x00\x00(\x00\x00\x00\x06\x00\x00\x004\x00\x00\x00\x01\x00\x00\x00}\x17\x06\x00\xb5t\x1d\x85\x14\x00\x00\x00\x14\x00\x00\x00E\x00\x00\x14\x00\x01\x00\x00@\x00|\xe7\x7f\x00\x00\x01\x7f\x00\x00\x014\x00\x00\x00")
>>> rdpcap(file)
<No name: TCP:0 UDP:0 ICMP:0 Other:1>
>>> 
>>> file = BytesIO(b'\n\r\r\n\xff\xfe\xfe\xffM<+\x1a')
>>> try:
...     rdpcap(file)
... except Scapy_Exception:
...     pass
... 
PcapNg: Could not read major value
>>> file = BytesIO(b"\n\r\r\n\x00\x00\x008\x1a+<M\x00\x01\x00\x00\xef\xff\xff\x81\x01\x00\x00\x01\x00\x00\x00\x0c\x00\x00\x00\x00\x00\x00\x00\x009\x00\x00\x00\x01\x00\x00\x000\x00e\x00\x00\x00\x00\x00\x00\x00\x008\x00\x00\x00\x06\x00\x00\x00d\x00\x00\x00\x00'\x00\x00\x00\x01\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x00\x80\x01\x00\x00\x000\x00\x00\x00\x01\x00\x00\x008\x00\x00\n\x0c\x00A\x05\xc0\x01\x00\x00\x00\x00\x00d\x00\x00\x00\x00\x00\x00\x00\x00\x000\x00\x00\x00\x01\x00\x00\x008\x00\x00\n\x0c\x00A\x05\xc0\x83\x00\x00\x043\x01\x00\x00\x00\x00\x00d\x00\x00\x00\x00d")
>>> l = rdpcap(file)
PcapNg: invalid option length 12 for end-of-option
PcapNg: EPB invalid proc index(expected 4 bytes, got 0) !
>>> 
>>> file = BytesIO(b"M<\xb2\xa1 \x00\xff\xf9\xff\x00\x00\xff\xff\xff\xa1\xb2<M\x01\x00\x00\x00\x00\x00\x00\x17\xca\xca\xca\xca\xca\xca\xca\xca\xca\xca\r\x80\x01[\x00\x01noenum\x00\x00\x00\x00\x00\x06value\x01\x00\x00\x00\x00\x00\x00\x01\x85\xff\xff\x00\x00\x94\xc3\x000\x00\x08\xa0\xf30\x800\x000\x00")
>>> l = rdpcap(file)
Traceback (most recent call last):
  File "<input>", line 2, in <module>
  File "/builddir/build/BUILD/scapy-2.5.0/scapy/utils.py", line 1346, in rdpcap
    return fdesc.read_all(count=count)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/builddir/build/BUILD/scapy-2.5.0/scapy/utils.py", line 1627, in read_all
    res = self._read_all(count)
          ^^^^^^^^^^^^^^^^^^^^^
  File "/builddir/build/BUILD/scapy-2.5.0/scapy/utils.py", line 1535, in _read_all
    p = self.read_packet()  # type: Packet
        ^^^^^^^^^^^^^^^^^^
  File "/builddir/build/BUILD/scapy-2.5.0/scapy/utils.py", line 1591, in read_packet
    rp = super(PcapReader, self)._read_packet(size=size)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/builddir/build/BUILD/scapy-2.5.0/scapy/utils.py", line 1503, in _read_packet
    return (self.f.read(caplen)[:size],
            ^^^^^^^^^^^^^^^^^^^
OverflowError: cannot fit 'int' into an index-sized integer
```

### Scapy version

1935723c

### Python version

3.12.1

### Operating system

Fedora 40

### Additional environment information

_No response_

### How to reproduce

It can be reproduced by running the test on an i386 machine.

### Actual result

_No response_

### Expected result

_No response_

### Related resources

_No response_

### PR DESCRIPTION
**Pull request by guedou**
Fixes #4527.

### CODE REVIEW

File: scapy/utils.py [hunk 1]
```diff
@@ -1500,7 +1500,14 @@ def _read_packet(self, size=MTU):
         if len(hdr) < 16:
             raise EOFError
         sec, usec, caplen, wirelen = struct.unpack(self.endian + "IIII", hdr)
-        return (self.f.read(caplen)[:size],
+
+        try:
+            data = self.f.read(caplen)[:size]
+        except OverflowError as e:
+            warning(f"Pcap: {e}")
```
Comment (evverx) [inline scapy/utils.py:1507]: I have to admit I haven't figured out where the overflow comes from yet but it seems that that issue is probably unrelated to the OSS-Fuzz crash. I think the test itself can be "fixed" by replacing the OSS-Fuzz testcase with something like
```sh
b'\xd4\xc3\xb2\xa1\x02\x00\x04\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff\x00\x00\x01\x00\x00\x00\x04{\xdcf\xc2\xa5\x07\x008\x00\x00\x008\x00\x00\x00A]+\xdb]\x04\x8e(6\n\x99\xcb\x08\x00E\x00\x00*\x00\x01\x00\x00@\x06\xe3V\x07\x87\xa5m\x17\x15\xd3m\x01\x85\x01\x85\x00\x00\x00\x00\x00\x00\x00\x00P\x02 \x00\xc5_\x00\x000\x00'
```
it triggers the crash without 1935723c and doesn't trigger the overflow. The overflow can probably be looked into separately with a separate testcase without blocking the next release (as far as I understand it's coming) and distros running the test suite on those platforms.

Comment (evverx) [reply scapy/utils.py:1507]: (I opened https://github.com/secdev/scapy/pull/4530 with the test case)

**Review by polybassa — APPROVED** [review]

**Comment by guedou** [comment]
My understanding is that caplen is a 32bits unsigned int, while the read parameter is a 32bits signed int. Hence, high values don’t fit.Sent from my iPhoneOn 7 Sep 2024, at 21:20, Nils Weiss ***@***.***> wrote:﻿
@polybassa approved this pull request.

**Comment by evverx** [comment]
As far as I can see in this particular case the length of the packet is `3402287818` and it's much more than `sys.maxsize` on i386 (`2147483647`). If `size` in `self.f.read(caplen)[:size]` is reasonable it should be possible to call `read` several times with some reasonable value to get the first bytes and skip the rest without triggering the overflow.

(Then again I'm not sure how important that is)

**Comment by evverx** [comment]
I took a look at what Wireshark does. Looks like it doesn't read more than 262144 bytes by default:
https://github.com/wireshark/wireshark/blob/4a2965cad39332f47b472fb873817ed021bffccb/wiretap/wtap.h#L363-L367
```
 * We don't want to write out files that specify a maximum packet size
 * greater than 262144 if we don't have to, as software reading those
 * files might allocate a buffer much larger than necessary, wasting memory.
 */
#define WTAP_MAX_PACKET_SIZE_STANDARD    262144U
```
and that pcap file is rejected with
```sh
(pcap: File has 3402287818-byte packet, bigger than maximum of 262144)
```
I think in practice it should be OK to catch the overflow and bail out then. Though I think it should be a separate test case.

**Comment by guedou** [comment]
2147483647 will be a 2Gb packet, that why I assumed that it was OK to only catch the OverflowError exception =)

**Comment by guedou** [comment]
If this MR is OK to you, let's merge it, then we will apply #4530

**Comment by gpotter2** [comment]
Bah it's the opposite way. First we merge #4530 then you come up with a different unit test in this PR that only tests this issue :) (or reuse the previous one I guess).

**Comment by guedou** [comment]
What do you want to change in the current test?

**Comment by evverx** [comment]
> What do you want to change in the current test?

I think something like
```python
b'\xd4\xc3\xb2\xa1\x02\x00\x04\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff\x00\x00\x01\x00\x00\x00%\xa8\xddfK\x1b\x05\x00\xca\xca\xca\xca*\x00\x00\x00\xff\xff\xff\xff\xff\xff\x86"\x11&\xab3\x08\x06\x00\x01\x08\x00\x06\x04\x00\x01]\x80\x0f\x13*r\n\x00\x02\x0f\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
```
should work. It contains `\xca\xca\xca\xca` leading to
```
OverflowError: cannot fit 'int' into an index-sized integer
```
on i386. On 64-bit machines scapy should read a normal ARP packet without triggering any other issues:
```
0000 Ether / ARP who has 0.0.0.0 says 10.0.2.15
```
I also passed that pcap to tshark and it rejects it with
```
(pcap: File has 3402287818-byte packet, bigger than maximum of 262144)
```
as expected.

**Comment by guedou** [comment]
Is there really a need for two tests that will each give different results on 32 and 64 bits platforms?Currently, the file from oss-fuzz allows to test the overflow issue on 32 bits platforms, and the LDAP files access on 64 bits ones.Sent from my iPhoneOn 8 Sep 2024, at 16:02, Evgeny Vereshchagin ***@***.***> wrote:﻿

What do you want to change in the current test?

I think something like
b'\xd4\xc3\xb2\xa1\x02\x00\x04\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff\x00\x00\x01\x00\x00\x00%\xa8\xddfK\x1b\x05\x00\xca\xca\xca\xca*\x00\x00\x00\xff\xff\xff\xff\xff\xff\x86"\x11&\xab3\x08\x06\x00\x01\x08\x00\x06\x04\x00\x01]\x80\x0f\x13*r\n\x00\x02\x0f\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
should work. It contains \xca\xca\xca\xca leading to
OverflowError: cannot fit 'int' into an index-sized integer

on i386. On 64-bit machines scapy should read a normal ARP packet without triggering any other issues:
0000 Ether / ARP who has 0.0.0.0 says 10.0.2.15

I also passed that pcap to tshark and it rejects it with
(pcap: File has 3402287818-byte packet, bigger than maximum of 262144)

as expected.

**Comment by evverx** [comment]
> Is there really a need for two tests that will each give different results on 32 and 64 bits platforms?

My thinking was that since the original testcase triggers the overflow (apart from the LDAP crash) it prevents the LDAP crash from being tested on those platforms. To test the LDAP crash and the overflow there ideally there should be two different testcases.

**Comment by guedou** [comment]
It still don’t see why we need to test the LDAP crash on a specific platform, but as you can provided a working test case, let’s use it =)Sent from my iPhoneOn 8 Sep 2024, at 17:48, Evgeny Vereshchagin ***@***.***> wrote:﻿

Is there really a need for two tests that will each give different results on 32 and 64 bits platforms?

My thinking was that since the original testcase triggers the overflow (apart from the LDAP crash) it prevents the LDAP crash from being tested on those platforms. To test the LDAP crash and the overflow there ideally there should be two different testcases.

**Comment by evverx** [comment]
> It still don’t see why we need to test the LDAP crash on a specific platform

I agree that it isn't i386-specific but if nothing else coverage reports would look better with those two lines covered there as well :-) When I opened https://github.com/secdev/scapy/pull/4530 I hadn't figured out what the issue was so to me it seemed reasonable to unblock the test on 32-bit machines while the overflow issue was in the process of being fixed.

**Comment by guedou** [comment]
I pushed the alternative test.

**Comment by evverx** [comment]
I ran it on i386 and it passed
```python
# python3 -m scapy.tools.UTscapy -t test/regression.uts -n 167
...
>>> l = rdpcap(file)
Pcap: cannot fit 'int' into an index-sized integer
>>> assert len(l) == 0 or ARP in l[0]
```

(what's interesting is that it turns out Wireshark rejects snapshot lengths larger than 2147483647 on 64 bit platforms too even though those in theory should be unsigned integers)

**Review by polybassa — APPROVED** [review]

**Pull request merged**